### ISSUE TITLE
Two-node DDP ResNet-50/Adam run ends with matching rank-1 loss but gradient and parameter-fingerprint divergence across CPU/Gloo and CUDA/NCCL

### ISSUE THREAD

**Issue by whitephilomel** [pytorch/pytorch#187276]
### 🐛 Describe the bug

This case is different from a simple final-loss mismatch. In a two-node `DistributedDataParallel` run, CPU/Gloo and CUDA/NCCL finish the same deterministic FP32 ResNet-50 workload with:

- nearly identical rank-1 final loss;
- very different final gradient norm;
- very different final parameter fingerprint;
- different final logits norm.

Important runtime settings:

```text
model                         = resnet50
optimizer                     = Adam
learning rate                 = 0.1
weight decay                  = 0.0001
scheduler                     = step
precision                     = fp32
AMP                           = disabled
TF32                          = disabled
gradient clipping             = 1.0
per-rank batch size           = 4
world size                    = 2
steps                         = 20
seed                          = 1
DDP broadcast_buffers         = true
DDP bucket_cap_mb             = 25
DDP gradient_as_bucket_view   = true
DDP static_graph              = false
```
## Reproduction

Attach these two files:

```text
minimal_repro_ddp_differential.py
pytorch_official_repro_config.json
```
The repro code is here: https://gist.github.com/whitephilomel/de633765555fbeaeb311ed3fc8b7bfce

## Actual behavior

The latest confirmed run with `CUBLAS_WORKSPACE_CONFIG=:4096:8` produced:

| rank | CPU/Gloo loss | CUDA/NCCL loss | loss rel diff | CPU grad norm | CUDA grad norm | grad rel diff | CPU param sum | CUDA param sum | param-sum rel diff |
|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| 0 | 4.346704959869385 | 3.0898776054382324 | 28.914% | 7.804810047149658 | 0.7611870169639587 | 90.247% | -121450.7109375 | -11323.53125 | 90.676% |
| 1 | 2.1507763862609863 | 2.155442953109741 | 0.2165% | 7.804810047149658 | 0.7611870169639587 | 90.247% | -121450.7109375 | -11323.53125 | 90.676% |

Rank 1 is the clearest symptom: the final loss is almost identical, but the final gradient norm and parameter fingerprint are not.

The summaries emitted by the two ranks were:

```json
{
  "rank": 0,
  "status": "DIVERGENCE_REPRODUCED",
  "reason": "arms diverged: final_loss, final_grad_norm, param_sum",
  "final_loss": {
    "cpu": 4.346704959869385,
    "gpu": 3.0898776054382324,
    "abs_diff": 1.2568273544311523,
    "rel_diff": 0.2891448501876049
  },
  "final_grad_norm": {
    "cpu": 7.804810047149658,
    "gpu": 0.7611870169639587,
    "abs_diff": 7.0436230301856995,
    "rel_diff": 0.9024720637189695
  },
  "param_sum": {
    "cpu": -121450.7109375,
    "gpu": -11323.53125,
    "abs_diff": 110127.1796875,
    "rel_diff": 0.9067643889229497
  }
}
```

```json
{
  "rank": 1,
  "status": "DIVERGENCE_REPRODUCED",
  "reason": "arms diverged: final_grad_norm, logits_norm, param_sum",
  "final_loss": {
    "cpu": 2.1507763862609863,
    "gpu": 2.155442953109741,
    "abs_diff": 0.004666566848754883,
    "rel_diff": 0.002165015242932895
  },
  "final_grad_norm": {
    "cpu": 7.804810047149658,
    "gpu": 0.7611870169639587,
    "abs_diff": 7.0436230301856995,
    "rel_diff": 0.9024720637189695
  },
  "param_sum": {
    "cpu": -121450.7109375,
    "gpu": -11323.53125,
    "abs_diff": 110127.1796875,
    "rel_diff": 0.9067643889229497
  }
}
```

Both ranks returned exit code `0`.

## Expected behavior

The CPU/Gloo and CUDA/NCCL arms do not need bitwise-identical results. However, under the same two-node DDP topology, fixed seed, deterministic input tensors, FP32 execution, and 20 short optimizer steps, I would not expect the
final loss on rank 1 to remain almost unchanged while the gradient norm and parameter fingerprint diverge by about 90%.

If this is expected numerical drift for this DDP/backend/device combination, it would be helpful to document that the final scalar loss may remain close even when the underlying gradient and parameter state diverges significantly.

### Versions

PyTorch: 2.9.1+cu128
Python: 3.11.15
CUDA reported by PyTorch: 12.8
NCCL: 2.27.5+cuda12.9
OS: Linux 5.15.0-113-generic x86_64, glibc 2.35
GPU: Tesla V100S-PCIE-32GB, one GPU per machine
Distributed topology: two machines, world_size=2, one process per machine
CPU：Intel(R) Xeon(R) Gold 6248R CPU @ 3.00GHz
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy

**Comment by arkadip-maitra** [comment]
this also seem like cpu gpu drift, even present on single device, not ddp/nccl issue
